libs/pageserver_api/src/models.rs

```diff
@@ -406,6 +406,13 @@ pub struct TenantLocationConfigRequest {
     pub config: LocationConfig, // as we have a flattened field, we should reject all unknown fields in it
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct TenantTimeTravelRequest {
+    pub current_shard_count: ShardCount,
```

Comment:
So I guess the idea of having multiple shard counts here is to enable time travel for a tenant that has undergone a split, so that we're recovering objects from the ancestral shard counts (which may still be referenced by the post-split shards).

Two thoughts:
- Specifying the shard count should probably be optional here: the storage controller can use the current shard count of the tenant as a sensible default
- Tenants can split more than once, so if the caller really wants to recover all ancestral shard counts of the tenant, they'd need to be able to specify a vector of shard counts, rather than just two.

Reply:
> the storage controller can use the current shard count of the tenant as a sensible default

Does the storage controller store the shard count for tenants that are detached? The tenant being detached is a hard requirement for usage of this API -- we don't want there to be any writes happening concurrent to S3 recovery.

> if the caller really wants to recover all ancestral shard counts of the tenant, they'd need to be able to specify a vector of shard counts, rather than just two.

Between the target time we want to revert to and the current time there is indeed the potential for a history of multiple shard counts. Say there were 4 shards at the start, and then it got split twice to 8 and 16 shards. Now we want to recover some point in time when there were 4 shards. So it has written to the shard prefixes for 8 shards, but assuming that shard splits eventually delete the prefixes for the prior shard number, then the 8 doesn't matter: we want to resurrect the state for 4 shards, and delete all the prefixes for 16 shards. We only have to invoke the lower level API for the 4 and 16 shards. In fact, we could special case the deletion for the 16 shards to do deletions directly, but that's an optimization that I'd keep for later.

This of course assumes that the API is not being invoked while a shard split is in progress. In which case we'd actually need to execute the recovery process for both the pre-split and post-split shard count, same goes when the target time is from the middle of a split.

Reply:
> A tenant in the detached state is still stored in the storage controller (it has policy=Detached, and an empty IntentState).

and what if the tenant has been deleted completely since? we also want to support that use case.

Reply:
I think for now I'll just use a test